Create config-in-database.rst (#2845)
mattermost/docs: source/administration/config-in-database.rst

Configuration in the Mattermost Database
=========================================

> A new configuration database was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to allow using the database for the single source of truth of the active configuration of your Mattermost installation. This provides a convenient way to manage configuration changes from the system console and provides a failsafe in high availability deployments when a server in the node goes down.  Additionally, in high availability deployments, configuration changes are applied to all nodes instantaneously as possible removing the need to update individual configuration files on each server and the read-only state of the system console is removed.   

Reviewer: Do you mean `A new configuration option`

Reviewer: - `system console` -> `System Console`
- `failsafe` may not be obvious for non-English native spakers

Reviewer: Wondering if there's a simpler way to write the last sentence. Or maybe having the benefits in bullets helps with readability

Author: I have addressed this feedback.

Reviewer: There's one more `system console` -> `System Console` 🙏

Author: Thanks - missed that one.

Configuration in the Mattermost Database
=========================================
A new configuration option was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database for the single source of truth of the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration table in the database. Mattermost has been running our high availability community server on this option since the feature was released.  

Benefits to using this option include: 

>   - Conveniently manage configuration changes directly from the System Console.

Reviewer 2: ```suggestion
  - Conveniently manage configuration changes directly from the System Console, even in High Availability deployments and read-only containerized environments.
```

>   - In high availability deployments, configuration changes are applied to all nodes instantaneously as possible removing the need to update individual configuration files on each server and ensures all servers have the correct configurations in the event  a server in the node goes down.

Reviewer 2: ```suggestion
  - Ensure all servers in a High Availability deployment have the same configuration, even when new servers are added to the cluster.
  - Automatically deploy SAML certificates and keys to all servers in the cluster.
```

>   - Also in high availability deployments, the read-only state of the System Console is removed.   

Reviewer 2: Strictly speaking, I'd say this isn't a "benefit".

```suggestion
```

> To change your server from reading the config.json file to using the configuration in database: 

Reviewer 2: Can't make these suggestions all in place, but propose replacing as follows:

```
To start using configuration in database, pass the database connection string via the `--config` flag or `MM_CONFIG` environment variable. For example:

    ```
    ./mattermost --config="postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable\u0026connect_timeout=10"
    ```

To migrate an existing config.json into the database, use the ``config migrate`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-config-migrate>`_. For example:

    ```
    ./mattermost config migrate  path/to/config.json "postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable&connect_timeout=10"
    ```

Any existing SAML certificates and private keys will also be migrated to the database.
```

1. Use the ``--config`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to change the binary from reading from the default config.json file and location to the config database store by specifying a supported provider uri. Alternatively, the you may specify the ``MM_CONFIG`` variable to configure the same flag. 
  
  a. The provider uri specifies the driver name and data source name used to resolve the database. For example: ``postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable\u0026connect_timeout=10``
  b. Alternatively, you can  use the ``--migrate config`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-config-migrate>`_ for installs that are already using the ``config.json`` default file. 
  
2. Set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true to disable automatically watching this default config.json and effecting changes.

> Once you are using the configuration in the database, changes to configuration settings will create a new entry in the configuration database, storing a copy of the previous configuration as a json blob. 

Reviewer 2: ```suggestion
With configuration in the database enabled, any changes to the configuration will recorded the new, full configuration as a new row in the `Configurations` table. Note that `ClusterSettings.ReadOnlyConfig` is ignored when using configuration in the database, enabling full use of the system console.
```

> Any environment variable configurations will continue to override settings from config.json, triggering them to appear read-only in the system console. When reading the configuration in the database, ClusterSettings.ReadOnlyConfig setting is automatically disabled as the database configuration ignores the config read only flag. 

Reviewer 2: ```suggestion
Note that environment variable overrides remain fully supported, and will no longer be written back to the configuration when a different change is saved.
```

.. note::

>     If using SAML, ensure the SAML certificates and keys are accessible to also migrate into the database.

Reviewer 2: Consolidated into the above.

Configuration in the Mattermost Database
=========================================

> A new configuration option was added in `5.10 <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database for the single source of truth of the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration table in the database. 

Reviewer 3: ```suggestion
A new configuration option was added in `5.10 release <https://docs.mattermost.com/administration/changelog.html#configuration-in-database>`_ to use the database as the single source of truth for the active configuration of your Mattermost installation. This will change the Mattermost binary from reading the default config.json file to reading the configuration settings stored within a configuration table in the database. 
```

> Mattermost has been running our community server on this option since the feature was released and recommends using it to those in High Availability deployments.  

Reviewer: ```suggestion
Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends using it to those in :doc:`High Availability deployments <cluster>`.
```

> Benefits to using this option include: 

Reviewer: ```suggestion
Benefits to using this option:
```

> Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends using it to those in :doc:`High Availability deployments <cluster>`.

Reviewer 3: ```suggestion
Mattermost has been running our `community server <https://community.mattermost.com>`_ on this option since the feature was released, and recommends its use for those on :doc:`High Availability deployments <cluster>`.
```

  - Conveniently manage configuration changes directly from the System Console, even in High Availability deployments and read-only containerized environments.
  - Ensure all servers in a High Availability deployment have the same configuration, even when new servers are added to the cluster.
  - Automatically deploy SAML certificates and keys to all servers in the cluster.

> To start using configuration in database, pass the database connection string via the `--config` flag or `MM_CONFIG` environment variable. For example:

Reviewer: ```suggestion
To start using configuration in database, pass the database connection string via the ``--config`` flag or ``MM_CONFIG`` environment variable. For example:
```

    ```
    ./mattermost --config="postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable\u0026connect_timeout=10"
    ```

To migrate an existing config.json into the database, use the ``config migrate`` `command <https://docs.mattermost.com/administration/command-line-tools.html#mattermost-config-migrate>`_. For example:

>     ```

Reviewer: This needs the code block formatting like

```
  .. code-block:: 

    ./mattermost config migrate  path/to/config.json "postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable&connect_timeout=10"
```

Reviewer: (Same for the code block above this one)

    ./mattermost config migrate  path/to/config.json "postgres://mmuser:mostest@dockerhost:5432/mattermost_test?sslmode=disable&connect_timeout=10"
    ```

Any existing SAML certificates and private keys will also be migrated to the database.

> Addtitionally, to disable automatically watching the default config.json and effecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.

Reviewer: ```suggestion
Addtitionally, to disable automatically watching the default config.json and affecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.
```

> Addtitionally, to disable automatically watching the default config.json and affecting changes, set the ``--disableconfigwatch`` `flag <https://docs.mattermost.com/administration/command-line-tools.html#mattermost>`_ to true.

Reviewer 2: I don't think this paragraph belongs in this document -- the focus is on switching /to/ database in the config, and this flag does nothing in that configuration.
```suggestion
```

> With configuration in the database enabled, any changes to the configuration are recorded to the `Configurations` table. Furthermore, `ClusterSettings.ReadOnlyConfig` is ignored, enabling full use of the System Console.

Reviewer: ```suggestion
With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.
```

> With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` table. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.

Reviewer 2: With configuration in the database enabled, any changes to the configuration are recorded to the ``Configurations`` and ``ConfigurationFiles`` tables. Furthermore, ``ClusterSettings.ReadOnlyConfig`` is ignored, enabling full use of the System Console.

> Note that environment variable overrides remain fully supported using the configuration in the database, and will no longer be written back to the configuration when a change is saved, but will be written to the database instead.

Reviewer 3: ```suggestion
Note that environment variable overrides remain fully supported using the configuration in the database. They will no longer be written back to the configuration when a change is saved, but will be written to the database instead.
```

> Note that environment variable overrides remain fully supported using the configuration in the database. They will no longer be written back to the configuration when a change is saved, but will be written to the database instead.

Reviewer 2: In fact, they won't be written anywhere:
```
Note that environment variable overrides remain fully supported using the configuration in the database, allowing per-server customization as required. A long standing bug has also been fixed, ensuring environment variables are no longer written back to the configuration when other changes are made.
```
